feat(aws-glue): add glue connection L2 Constructs

wulfmann opened the pull request:
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

<!-- 
Please read the contribution guidelines and follow the pull-request checklist:
https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md
 -->

wulfmann commented:
@skinny85 this is ready for a review. I made the changes you suggested on #5171

wulfmann commented on packages/@aws-cdk/aws-glue/lib/connection.ts:
```diff
@@ -0,0 +1,114 @@
+  constructor(scope: Construct, id: string, props: ConnectionProps) {
+    super(scope, id);
+
+    const subnetId = props.vpcSubnets && props.vpc ? props.vpc.selectSubnets(props.vpcSubnets).subnetIds[0] : undefined;
```
@skinny85 I wasn't entirely sure what you meant on the other PR. Is this what you intended for the subnets? I have the same question for line 106.

skinny85 replied:
It should be:

```ts
const subnetId = props.vpc ? props.vpc.selectSubnets(props.vpcSubnets).subnetIds[0] : undefined;
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 69bb59b2efe57d6e7add708d78da050a3b7362e3
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f05c7da7-300d-4248-a26e-c7963bb6fbb7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 36c92e9d5b4c84ecbde4be8a74b7edca4981a488
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=57d15dfc-27c1-4802-9b9e-7c442f660d96%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-glue/lib/connection.ts:
```diff
@@ -0,0 +1,114 @@
+  readonly connectionName: string;
+}
+
+export interface ConnectionAttributes {
```
This interface is never used - feel free to remove for now (I assume you'e gearing up towards a `fromConnectionAttributes` method with it).

wulfmann replied:
removed for now

skinny85 commented on packages/@aws-cdk/aws-glue/lib/connection.ts:
```diff
@@ -0,0 +1,114 @@
+  /**
+   * The name of the connection
+   */
+  readonly connectionName: string;
```
Missing `@attribute` annotation.

wulfmann replied:
fixed

skinny85 commented on packages/@aws-cdk/aws-glue/lib/connection.ts:
```diff
@@ -0,0 +1,114 @@
+  readonly connectionName: string;
+}
+
+export interface ConnectionProps {
```
Missing docs (something simple like "Construction properties for {@link Connection}" is enough in this case).

wulfmann replied:
fixed

skinny85 commented on packages/@aws-cdk/aws-glue/lib/connection.ts:
```diff
@@ -0,0 +1,114 @@
+export interface ConnectionProps {
+  /**
+   * The name of the connection
+   * @default no name
```
I believe the default is a CloudFormation-generated name, is it not?

wulfmann replied:
fixed

skinny85 commented on packages/@aws-cdk/aws-glue/lib/connection.ts:
```diff
@@ -0,0 +1,114 @@
+   * but it will be deprecated in the future.
+   * @default no availability zone set
+   */
+  readonly availabilityZone?: string;
```
Can we get rid of this then, in favor of the `vpc` - `vpcSubnets` pair?

wulfmann replied:
fixed

skinny85 commented on packages/@aws-cdk/aws-glue/lib/connection.ts:
```diff
@@ -0,0 +1,114 @@
+   * The security group ID list used by the connection.
+   * @default no security group ids
+   */
+  readonly securityGroupIds?: ec2.ISecurityGroup[];
```
The name should be just `securityGroups`.

wulfmann replied:
fixed

skinny85 commented on packages/@aws-cdk/aws-glue/lib/connection.ts:
```diff
@@ -0,0 +1,114 @@
+  /**
+   * The vpc used by the connection
+   */
+  readonly vpc?: ec2.Vpc;
```
Is `vpc` really optional? What happens if you don't provide it?

wulfmann replied:
Yes as far as I can tell all of the physical connection requirements are optional: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-glue-connection-physicalconnectionrequirements.html

skinny85 commented on packages/@aws-cdk/aws-glue/lib/connection.ts:
```diff
@@ -0,0 +1,114 @@
+
+  /**
+   * The subnets used by the connection
+   * @default no subnets
```
Actually, the default is "private subnets", but that's when `vpc` is required...

wulfmann replied:
fixed

skinny85 commented on packages/@aws-cdk/aws-glue/lib/connection.ts:
```diff
@@ -0,0 +1,114 @@
+  public readonly connectionName: string;
+
+  constructor(scope: Construct, id: string, props: ConnectionProps) {
+    super(scope, id);
```
You want:

```suggestion
    super(scope, id, {
      physicalName: props.connectionName,
    });
```

wulfmann replied:
updated

skinny85 commented on packages/@aws-cdk/aws-glue/lib/connection.ts:
```diff
@@ -0,0 +1,114 @@
+        matchCriteria: props.matchCriteria,
+        name: props.connectionName,
+        physicalConnectionRequirements: {
+          availabilityZone: props.availabilityZone,
```
This should almost definitely be taken from the subnet acquired from doing `vpc.selectSubnets(vpcSubnets)`.

wulfmann replied:
fixed

skinny85 commented on packages/@aws-cdk/aws-glue/test/test.connection.ts:
```diff
@@ -0,0 +1,38 @@
+import * as glue from '../lib';
+import { Vpc } from '@aws-cdk/aws-ec2';
+
+export = {
```
Can you use Jest tests instead of Nodeunit? We're trying to get rid of Nodeunit in our codebase. Should be super easy, just add an (empty) `jest` block to your `package.json` (like this: `"jest": {}`), and then name your tests `connection.test.ts` instead.

wulfmann replied:
updated

skinny85 commented on packages/@aws-cdk/aws-glue/test/test.connection.ts:
```diff
@@ -0,0 +1,38 @@
+        connectionProperties: {},
+    });
+
+    expect(stack).toMatch({
```
I think you want `expect(stack).toHaveResourceLike(...)`.... right now, these tests don't actually verify anything.

wulfmann replied:
Yea im still working on tests.

skinny85 commented on packages/@aws-cdk/aws-glue/test/test.connection.ts:
```diff
@@ -0,0 +1,38 @@
+    });
+
+    test.done();
+  },
```
Empty line between tests (on the same level) please.

wulfmann replied:
fixed

skinny85 commented on packages/@aws-cdk/aws-glue/lib/connection.ts:
```diff
@@ -0,0 +1,114 @@
+import ec2 = require('@aws-cdk/aws-ec2');
```
We switched the import style in the meantime :).

```ts
import * as ec2 from '@aws-cdk/aws-ec2';
```

wulfmann replied:
fixed

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 6f380cebe671ef84211a6c47fe47693cc8528b77
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a8938205-3571-4a4d-b41b-a07ed1c14a66%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: cc31eb20486076d5c518f17ca819eac759e79130
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=91f11385-0bba-4c3c-98de-6f3c2ed651a3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8a383548ba48398628e30eeb385b1f2c6520c95c
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a4ba5e56-c949-4933-9bf1-c4c1d809b9ab%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 0590ff0ce78472d57f876c14d223aef03f58cef0
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=841c5b86-a55e-4d4f-9b7c-ff35c862eca8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 96df1df078ea551bd5b2f8a86479c7b6f1ef0e0b
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a9cede88-37ad-46cd-bbd3-c9e3b3ba8bc4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-glue/test/connection.test.ts:
```diff
@@ -0,0 +1,33 @@
+import * as glue from '../lib';
+import * as ec2 from '@aws-cdk/aws-ec2';
+
+export = {
```
Uhm. Maybe I did a mental shortcut saying "that's all you needed". Jest test look a tiny bit different than Nodeunit tests. So, this file would be something like:

```ts
describe('Glue Connection', () => {
  test('can be created without a VPC', () => {
    // ...
  });

  test('can be created with a VPC', () => {
    // ...
  });
});
```

skinny85 commented:
About the Jest tests - my apologies, I forgot this module already has some Nodeunit tests. In this case (because you can't mix Nodeunit and Jest in the same CDK package), just make the new tests in Nodeunit as well.

Apologies for the churn!

skinny85 commented:
BTW, we're investigating why the build fails with

```
 error TS0: Exported APIs cannot use un-exported type @aws-cdk/aws-glue.ConnectionInputTypes.JDBC
```

We know it's from the single-value enum, but it shouldn't really be happening. Will update when we know more.

skinny85 reviewed (dismissed):
Just removing from my to-do list.

skinny85 commented:
@wulfmann any updates here? Are you ready for another round of reviews, or still working on it? (The single-value enum thing should be resolved now)

wulfmann commented:
I will check this first thing tomorrow and see if we can get this resolved!

humanzz commented:
Hi everyone,
Just checking on the status of this as my team is really interested in `Job` constructs for glue.

humanzz commented:
Hi everyone,
Just checking on the status of this as my team is really interested in `Job` constructs for glue.

skinny85 commented:
Still waiting for @wulfmann on this one 🙂

humanzz commented on packages/@aws-cdk/aws-glue/lib/connection.ts:
```diff
@@ -0,0 +1,110 @@
+   * The type of the connection. Currently, only JDBC is supported;
+   * SFTP is not supported.
+   */
+  JDBC = 'JDBC',
```
Seems like there are more types currently supported https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-glue-connection-connectioninput.html

wulfmann replied:
updated

humanzz commented:
@skinny85 Any chance/way to help completing this PR if @wulfmann is not available to wrap it?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: a099a9bff47bd005f2775aed6543179f9496f3fe
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d1fac84a-2b2c-4a94-a396-511eabe91d08%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 2997ca5df7cb7cf1d8df579713abbdf6cb85426c
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1a9be57c-7364-4159-b6b3-797fbc78289f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

wulfmann commented:
Finally back to wrapping this up. Sorry to everyone waiting on this.

wulfmann commented:
I've updated to the current branch of upstream and cleaned up some of the missing things. I should be able to wrap the tests up tomorrow so we can at least get the L2 connection construct merged.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d017d0c5ec6d3ad2c2439f35d58e8a384b904a15
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bb7c59a8-c648-446b-a28a-55e972632669%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ee9041d84fa4a2630ec7bd63b279cdca7a02d8d2
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=baf3c187-e0b4-4d8d-835c-5b5f19e04fbd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c8cceef43a0583173ec4bab629e7d75bed8d69cc
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=528cae2d-d6ac-4517-89ee-d4f279e41a43%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b2bc5150b34a746894bd67b459274dde0e424e28
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=08b6c675-81f6-4428-b5c4-090c2a99d48f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 3201d7dbd67a9658e57440ed8423b20dba191b8a
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ae06074d-4a35-4b41-90fb-ea11db6073b7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 808c88ae963bf8a5aee56dc0965b7ac365c5407b
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ac21208c-1569-461a-8497-f39ba92b18c6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 45afa5b62158dde8b201442ece5f0aea10c5d44e
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=65558c69-e79a-43b5-823b-e8e4b73f75ad%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 reviewed (changes requested):
@wulfmann let me know when this is ready for review!

(Putting in "Request changes" to remove it from my ToDo list)

humanzz commented:
@wulfmann any updates here?

skinny85 commented:
I'm closing this one due to inactivity, please comment if you want it re-opened.
